### PR TITLE
Added chat page

### PR DESCRIPTION
**Pull request by bochka123**
image
I have issue with date of messages, but I will fix it in the morning with the fresh head
For now, pleas, just look at the code and tell me please if there is something more to fix, thanks)
Good night)

### CODE REVIEW

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 1]
```diff
@@ -1,19 +1,113 @@
...
+        <mat-form-field appearance="fill">
+            <input matInput placeholder="Type a name..."/>
+            <mat-icon matSuffix>search</mat-icon>
+        </mat-form-field>
+        <div class="line"></div>
+        <div class="people">
+            <div [className]="'person '
+                + (person.isOnline ? 'onlinePerson ' : 'offlinePerson ')
+                + (person.numberOfUnreadMessages > 0 ? 'unread ' : 'read ')
+                "
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:19]: Use https://angular.io/api/common/NgClass

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 2]
```diff
@@ -1,19 +1,113 @@
...
-            <button class="send-button" mat-raised-button color="primary" (click)="startSessionCall()">
-                Start
-            </button>
+<app-header></app-header>
+<main>
+    <aside>
+        <div class="message-header">
+            <span class="message-span">Messages</span>
+            <div class="total-number-of-messages">
+                <span>3</span>
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:7]: Can you make it dynamic?

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html (hunk 1 above)
Comment (Anatolii-Ivanov) [reply frontend/src/app/modules/chat/chat-page/chat-page.component.html:19]: Better to use it everywhere

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 3]
```diff
@@ -1,19 +1,113 @@
...
+            </div>
+        </header>
+        <div class="chat-body" *ngIf="messages">
+            <div *ngFor="let message of messages; index as i">
+                <div class="date" *ngIf="
+                i+1 !== messages.length &&
+                !(messages[i+1].createdAt.getFullYear() !== messages[i].createdAt.getFullYear() ||
+                messages[i+1].createdAt.getMonth() !== messages[i].createdAt.getMonth() ||
+                messages[i+1].createdAt.getDate() !== messages[i].createdAt.getDate()
+                )
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:67]: Move to ts and check via moment()

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 4]
```diff
@@ -1,19 +1,113 @@
...
+                ">
+                    <div class="date-line"></div>
+                    <div class="day">
+                        {{message.createdAt.getUTCDate()}}
+                        {{getMonthName(message.createdAt.getMonth()).slice(0, 3)}}
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:72]: Use DatePipe https://angular.io/api/common/DatePipe

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html (hunk 3 above)
Comment (Anatolii-Ivanov) [reply frontend/src/app/modules/chat/chat-page/chat-page.component.html:67]: Better to check all dates with it

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 5]
```diff
@@ -1,19 +1,113 @@
...
+                    </div>
+                </div>
+                <div class="message-photo"
+                     [className]="'message-photo ' + (message.userId === currentUserId ? 'own-message' : 'not-own-message')">
+                    <img src="assets/header/avatar.svg" *ngIf="message.userId !== currentUserId">
+                    <div class="message" [ngClass]="{ sender: message.userId === currentUserId }">
+                        <p>{{ message.text }}</p>
+                        <div class="message-time">
+                            <span>
+                                {{message.createdAt.getHours() < 10 ? '0'+message.createdAt.getHours() : message.createdAt.getHours()}}
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:83]: DatePipe

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

**Comment by Anatolii-Ivanov** [comment]
Concerning screen - please, check font styles on the right side, as they look different from design
And make one line messages centered in block vertically, please

**Comment by bochka123** [comment]
image
image

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 6]
```diff
@@ -1,19 +1,116 @@
...
+                    <div class="date"
+                         *ngIf="!i || message.createdAt.toDateString() !== lastDate.toDateString()"
+                         >
+                        <div class="date-line"></div>
+                        <div class="day">
+                                {{
+                                    message.createdAt.getFullYear() === today.getFullYear() ?
+                                        (getDateText(message.createdAt)) :
+                                        (message.createdAt | date:'dd MMM, YYYY')
+                                }}
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:76]: Can you make it in ts?
And maybe move messages to a separate component

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 7]
```diff
@@ -1,19 +1,116 @@
...
+                        </div>
+                        <ng-container *ngIf="!i || message.createdAt.toDateString() !== lastDate.toDateString()">
+                            {{ setLastDate(message.createdAt) }}
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:79]: Please map data, so you won't set anything in html

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/chat/message-group/message-group.component.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        yesterday.setDate(today.getDate() - 1);
+
+        if (date.toDateString() === today.toDateString()) {
+            return 'Today';
+        }
+        if (date.toDateString() === yesterday.toDateString()) {
+            return 'Yesterday';
+        }
+
+        return date.toLocaleDateString('en-US', { day: 'numeric', month: 'short' });
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/message-group/message-group.component.ts:39]: Use moment please

File: frontend/src/app/modules/chat/chat-page/chat-page.component.ts [hunk 9]
```diff
@@ -1,51 +1,101 @@
...
+
+                return groupDate.getTime() === messageDate.getTime();
+            });
+
+            if (messageGroupIndex === -1) {
+                result.push({ date: messageDate, messages: [message] });
+            } else {
+                result[messageGroupIndex].messages.push(message);
+            }
+        });
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.ts:70]: Use reduce()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce
And moment for comparing dates

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/chat/message-group/message-group.component.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+    showDate: boolean;
+
+    getDateText(date: Date): string {
+        const today = new Date();
+        const yesterday = new Date(today);
+
+        yesterday.setDate(today.getDate() - 1);
+
+        if (moment(date).isSame(moment(today).startOf('day'))) {
```
Comment (Anatolii-Ivanov) [inline frontend/src/app/modules/chat/message-group/message-group.component.ts:33]: moment(today) is the same as moment()

moment().addDays('days', -1) is the same as moment(yesterday)

**Review by Anatolii-Ivanov — APPROVED** [review]

File: frontend/src/app/modules/chat/chat-page/chat-page.component.html [hunk 11]
```diff
@@ -1,19 +1,87 @@
...
+            <mat-icon matSuffix>search</mat-icon>
+        </mat-form-field>
+        <div class="line"></div>
+        <div class="people">
+            <div [ngClass]="{
+                  'person': true,
+                  'onlinePerson': person.isOnline,
+                  'offlinePerson': !person.isOnline,
+                  'unread': person.numberOfUnreadMessages > 0,
+                  'read': person.numberOfUnreadMessages === 0
```
Comment (tatianahutii) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.html:21]:
```suggestion
                  'unread': person.numberOfUnreadMessages,
                  'read': !person.numberOfUnreadMessages
```

File: frontend/src/app/modules/chat/chat-page/chat-page.component.ts [hunk 12]
```diff
@@ -1,51 +1,100 @@
...
-
-        this.messages.push({
-            chatId: this.currentChatId,
-            userId: this.currentUserId,
-            text: message,
-            createdAt: new Date(),
-        });
+        if (message) {
+            this.form.reset();
+            this.messages.push({
```
Comment (tatianahutii) [inline frontend/src/app/modules/chat/chat-page/chat-page.component.ts:88]:
```suggestion
            this.messages. = [ ...messages, {
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by tatianahutii — DISMISSED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by ShevchenkoKirilo — APPROVED** [review]

**Pull request merged**